src/main/java/com/google/devtools/build/lib/runtime/BlazeRuntime.java

```diff
@@ -310,6 +312,44 @@ private BuildEventArtifactUploader newUploader(
     return buildEventArtifactUploaderFactoryMap.select(buildEventUploadStrategy).create(env);
   }
 
+  /**
+   * Implements automatic profile management.
+   *
+   * <ul>
+   *   <li>computes the path to write the profile for the current command to
+   *   <li>points a symlink called <code>"command.profile.gz"</code> to the current command's
+   *       profile
+   *   <li>does garbage collection for profiles from previous commands based on the <code>
+   * --profiles_to_retain</code> flag
+   * </ul>
+   *
+   * @return the path this command's profile should be written to
+   */
+  @VisibleForTesting
+  static Path manageProfiles(Path dir, String commandId, int retentionWindow) throws IOException {
+    var prefix = "command-";
+    var suffix = ".profile.gz";
+    var symlinkName = "command.profile.gz";
+    record PathAndMtime(Path path, long mtime) {}
+    var old = new ArrayList<PathAndMtime>();
+    for (var dirent : dir.readdir(Symlinks.FOLLOW)) {
+      if (symlinkName.equals(dirent.getName())) {
+        dir.getChild(dirent.getName()).delete();
+      } else if (dirent.getName().startsWith(prefix) && dirent.getName().endsWith(suffix)) {
+        var path = dir.getChild(dirent.getName());
+        old.add(new PathAndMtime(path, path.stat().getLastModifiedTime()));
+      }
+    }
+    old.sort(Comparator.comparingLong(PathAndMtime::mtime));
+    int toRemove = (old.size() >= retentionWindow) ? old.size() - retentionWindow + 1 : 0;
+    for (var i = 0; i < toRemove; i++) {
+      old.get(i).path().delete();
+    }
+    var profileName = prefix + commandId + suffix;
+    dir.getChild(symlinkName).createSymbolicLink(PathFragment.create(profileName));
```

Comment:
This will create a dangling junction on Windows as the target doesn't exist. Could you move this to the end of the build so that it can copy (or move) the completed profile instead?

Reply:
Changed to do `createSymbolicLink` at the end, though that makes everything less self-contained.